File: backend/src/main/java/be/lutske/leolegacy/interfaceadapter/rest/RecipeResource.java
package be.lutske.leolegacy.interfaceadapter.rest;

import be.lutske.leolegacy.application.usecase.GetRecipeUseCase;
import be.lutske.leolegacy.application.usecase.ListRecipesUseCase;
import be.lutske.leolegacy.application.usecase.ListRecipesByCategoryUseCase;
import be.lutske.leolegacy.application.usecase.IncrementRecipeViewCountUseCase;
import be.lutske.leolegacy.domain.recipe.Recipe;
import be.lutske.leolegacy.domain.recipe.RecipeId;
import be.lutske.leolegacy.domain.category.CategoryId;
import jakarta.inject.Inject;
import jakarta.ws.rs.*;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import java.util.List;

@Path("/api/recipes")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class RecipeResource {

    @Inject
    GetRecipeUseCase getRecipeUseCase;

    @Inject
    ListRecipesUseCase listRecipesUseCase;

    @Inject
    ListRecipesByCategoryUseCase listRecipesByCategoryUseCase;

    @Inject
    IncrementRecipeViewCountUseCase incrementViewCountUseCase;

    @GET
    public Response listRecipes(@QueryParam("categoryId") String categoryId) {
        if (categoryId != null) {
            CategoryId categoryIdObj = new CategoryId(categoryId);
            List<Recipe> recipes = listRecipesByCategoryUseCase.execute(categoryIdObj);
        return Response.ok(recipes).build();
    }
    else {
        List<Recipe> recipes = listRecipesUseCase.execute();
        return Response.ok(recipes).build();
    }

    }

    @GET
    @Path("/top")
    public Response topRecipes() {
        // Assuming this would be handled by a different use case if needed
        // For now, we'll keep it as is but note that this should be refactored
        return Response.status(Response.Status.NOT_IMPLEMENTED).build();
    }

    @GET
    @Path("/{id}")
    public Response getRecipe(@PathParam("id") String id) {
        RecipeId recipeId = new RecipeId(id);
        Recipe recipe = getRecipeUseCase.execute(recipeId);
        incrementViewCountUseCase.execute(recipeId);
        return Response.ok(recipe).build();
    }
}

File: backend/src/main/kotlin/be/lutske/leolegacy/interfaceadapter/rest/RecipeResource.kt
package be.lutske.leolegacy.interfaceadapter.rest

import be.lutske.leolegacy.infrastructure.persistence.entity.RecipeEntity
import be.lutske.leolegacy.infrastructure.persistence.repository.RecipeRepository
import jakarta.ws.rs.GET
import jakarta.ws.rs.NotFoundException
import jakarta.ws.rs.Path
import jakarta.ws.rs.PathParam
import jakarta.ws.rs.Produces
import jakarta.ws.rs.QueryParam
import jakarta.ws.rs.core.MediaType

/**
 * REST resource for recipes.
 */
@Path("/api/recipes")
class RecipeResource(
    private val recipeRepository: RecipeRepository
) {

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    fun listRecipes(@QueryParam("categoryId") categoryId: Long?): List<RecipeSummaryResponse> {
        val recipes: List<RecipeEntity> = if (categoryId != null) {
            recipeRepository.findByCategoryId(categoryId)
        } else {
            recipeRepository.listAll()
        }
        return recipes.map { it.toSummary() }
    }

    @GET
    @Path("/top")
    @Produces(MediaType.APPLICATION_JSON)
    fun topRecipes(): List<RecipeSummaryResponse> {
        return recipeRepository.findTopByViewCount(10).map { it.toSummary() }
    }

    @GET
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    fun getRecipe(@PathParam("id") id: Long): RecipeDetailResponse {
        val recipe = recipeRepository.findById(id)
            ?: throw NotFoundException("Recipe with id $id not found")
        return recipe.toDetail()
    }

    private fun RecipeEntity.toSummary() = RecipeSummaryResponse(
        id = this.id,
        title = this.title,
        categoryName = this.category.name,
        viewCount = this.viewCount
    )

    private fun RecipeEntity.toDetail() = RecipeDetailResponse(
        id = this.id,
        title = this.title,
        ingredients = this.ingredients.split("|").filter { it.isNotBlank() },
        preparation = this.preparation,
        categoryId = this.category.id,
        categoryName = this.category.name,
        viewCount = this.viewCount
    )
}
